pubsub/consumer.go

```diff
@@ -0,0 +1,152 @@
+	if cfg.RedisURL == "" {
+		return nil, fmt.Errorf("redis url cannot be empty")
+	}
+	c, err := redisutil.RedisClientFromURL(cfg.RedisURL)
+	if err != nil {
+		return nil, err
+	}
+	consumer := &Consumer[T]{
+		id:     uuid.NewString(),
+		client: c,
+		cfg:    cfg,
+	}
+	return consumer, nil
+}
+
+// Start starts the consumer to iteratively perform heartbeat in configured intervals.
+func (c *Consumer[T]) Start(ctx context.Context) {
+	c.StopWaiter.Start(ctx, c)
+	c.StopWaiter.CallIteratively(
+		func(ctx context.Context) time.Duration {
+			c.heartBeat(ctx)
+			return c.cfg.KeepAliveInterval
+		},
+	)
+}
+
+func (c *Consumer[T]) StopAndWait() {
+	c.StopWaiter.StopAndWait()
+}
+
+func heartBeatKey(id string) string {
+	return fmt.Sprintf("consumer:%s:heartbeat", id)
+}
+
+func (c *Consumer[T]) heartBeatKey() string {
+	return heartBeatKey(c.id)
+}
+
+// heartBeat updates the heartBeat key indicating aliveness.
+func (c *Consumer[T]) heartBeat(ctx context.Context) {
+	if err := c.client.Set(ctx, c.heartBeatKey(), time.Now().UnixMilli(), 2*c.cfg.KeepAliveTimeout).Err(); err != nil {
+		l := log.Info
+		if ctx.Err() != nil {
+			l = log.Error
+		}
+		l("Updating heardbeat", "consumer", c.id, "error", err)
+	}
+}
+
+// Consumer first checks it there exists pending message that is claimed by
+// unresponsive consumer, if not then reads from the stream.
+func (c *Consumer[T]) Consume(ctx context.Context) (*Message[T], error) {
+	res, err := c.client.XReadGroup(ctx, &redis.XReadGroupArgs{
+		Group:    c.cfg.RedisGroup,
+		Consumer: c.id,
+		// Receive only messages that were never delivered to any other consumer,
+		// that is, only new messages.
+		Streams: []string{c.cfg.RedisStream, ">"},
+		Count:   1,
+		Block:   time.Millisecond, // 0 seems to block the read instead of immediately returning
+	}).Result()
+	if errors.Is(err, redis.Nil) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("reading message for consumer: %q: %w", c.id, err)
+	}
+	if len(res) != 1 || len(res[0].Messages) != 1 {
+		return nil, fmt.Errorf("redis returned entries: %+v, for querying single message", res)
+	}
+	log.Debug(fmt.Sprintf("Consumer: %s consuming message: %s", c.id, res[0].Messages[0].ID))
+	var (
+		value = res[0].Messages[0].Values[messageKey]
+		tmp   T
+	)
+	val, err := tmp.Unmarshal(value)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshaling value: %v, error: %w", value, err)
+	}
+
+	return &Message[T]{
+		ID:    res[0].Messages[0].ID,
+		Value: val,
+	}, nil
+}
+
+func (c *Consumer[T]) ACK(ctx context.Context, messageID string) error {
+	log.Info("ACKing message", "consumer-id", c.id, "message-sid", messageID)
+	_, err := c.client.XAck(ctx, c.cfg.RedisStream, c.cfg.RedisGroup, messageID).Result()
+	return err
+}
+
+func (c *Consumer[T]) SetResult(ctx context.Context, messageID string, result string) error {
+	acquired, err := c.client.SetNX(ctx, messageID, result, c.cfg.KeepAliveTimeout).Result()
```

Comment:
I don't think the same timeout makes sense here. We don't want that to disappear if the node didn't test it for a long time